components/discount/freeRegisterCheckout.tsx

```diff
@@ -59,7 +95,110 @@ const FreeRegisterCheckout: FunctionComponent<FreeRegisterCheckoutProps> = ({
+  const [maxGasTokenAmount, setMaxGasTokenAmount] = useState<bigint>();
+  const [gasMethod, setGasMethod] = useState<GasMethod>("traditional");
   const { addTransaction } = useNotificationManager();
+  const { provider } = useProvider();
+
+  useEffect(() => {
+    setGasTokenPrice(gasTokenPrices[0]);
+  }, [gasTokenPrices]);
+
+  useEffect(() => {
+    if (gaslessCompatibility?.isCompatible && paymasterRewards.length > 0)
+      setGasMethod("paymaster");
+  }, [gaslessCompatibility, paymasterRewards]);
+
+  useEffect(() => {
+    if (!gaslessCompatibility?.isCompatible) setGasMethod("traditional");
+  }, [gaslessCompatibility]);
+
+  useEffect(() => {
+    fetchGaslessStatus(options).then((res) => {
+      setGaslessAPIAvailable(res.status);
+    });
+  }, []);
+
+  useEffect(() => {
+    if (gasMethod === "traditional" && loadingGas) setLoadingGas(false);
+  }, [gasMethod, loadingGas]);
+
+  useEffect(() => {
+    if (!account || !gaslessAPIAvailable) return;
+    fetchAccountCompatibility(account.address, options).then(
+      setGaslessCompatibility
+    );
+    fetchAccountsRewards(account.address, {
+      ...options,
+      protocol: "gasless-sdk",
+    }).then(setPaymasterRewards);
+  }, [account, gaslessAPIAvailable]);
+
+  const estimateCalls = useCallback(
+    async (
+      account: AccountInterface,
+      calls: Call[]
+    ): Promise<EstimateFeeResponse> => {
+      const contractVersion = await provider.getContractVersion(
+        account.address
+      );
+      const nonce = await provider.getNonceForAddress(account.address);
+      const details = stark.v3Details({ skipValidate: true });
+      const invocation = {
+        ...details,
+        contractAddress: account.address,
+        calldata: transaction.getExecuteCalldata(calls, contractVersion.cairo),
+        signature: [],
+      };
+      return provider.getInvokeEstimateFee(
+        { ...invocation },
+        { ...details, nonce },
+        "pending",
+        true
+      );
+    },
+    [provider]
+  );
+
+  useEffect(() => {
+    fetchGasTokenPrices(options).then(setGasTokenPrices);
+  }, []);
+
+  useEffect(() => {
+    if (
+      !account ||
+      !gasTokenPrice ||
+      !gaslessCompatibility ||
+      !gaslessAPIAvailable ||
+      gasMethod === "traditional"
+    )
+      return;
+    setLoadingGas(true);
+    estimateCalls(account, callData).then((fees) => {
+      const estimatedGasFeesInGasToken = getGasFeesInGasToken(
+        BigInt(fees.overall_fee),
+        gasTokenPrice,
+        BigInt(fees.gas_price),
+        BigInt(fees.data_gas_price ?? "0x1"),
+        gaslessCompatibility.gasConsumedOverhead,
+        gaslessCompatibility.dataGasConsumedOverhead
+      );
+      setMaxGasTokenAmount(estimatedGasFeesInGasToken * BigInt(2));
+      setLoadingGas(false);
+    });
+  }, [
+    gasMethod,
+    callData,
+    account,
+    gasTokenPrice,
+    gaslessCompatibility,
+    estimateCalls,
+    gaslessAPIAvailable,
+  ]);
 
```

Comment:
Seems big chunk of code. 

Could you make it a separate hook especially if we can reuse this elsewhere in another registration flow ...